kernel/src/table_changes/scan_file.rs

```diff
@@ -0,0 +1,414 @@
+    /// the same path as this [`CdfScanFile`]
+    pub remove_dv: Option<DvInfo>,
+    /// A `HashMap<String, String>` which are partition values
+    pub partition_values: HashMap<String, String>,
+    /// The commit version that this action was performed in
+    pub commit_version: i64,
+    /// The timestamp of the commit that this action was performed in
+    pub commit_timestamp: i64,
+}
+
+pub(crate) type CdfScanCallback<T> = fn(context: &mut T, scan_file: CdfScanFile);
+
+/// Transforms an iterator of TableChangesScanData into an iterator  of
+/// `UnresolvedCdfScanFile` by visiting the engine data.
+#[allow(unused)]
+pub(crate) fn scan_data_to_scan_file(
+    scan_data: impl Iterator<Item = DeltaResult<TableChangesScanData>>,
+) -> impl Iterator<Item = DeltaResult<CdfScanFile>> {
+    scan_data
+        .map(|scan_data| -> DeltaResult<_> {
+            let scan_data = scan_data?;
+            let callback: CdfScanCallback<Vec<CdfScanFile>> =
+                |context, scan_file| context.push(scan_file);
+            Ok(visit_cdf_scan_files(&scan_data, vec![], callback)?.into_iter())
+        }) // Iterator-Result-Iterator
+        .flatten_ok() // Iterator-Result
+}
+
+/// Request that the kernel call a callback on each valid file that needs to be read for the
+/// scan.
+///
+/// The arguments to the callback are:
+/// * `context`: an `&mut context` argument. this can be anything that engine needs to pass through to each call
+/// * `CdfScanFile`: a [`CdfScanFile`] struct that holds all the metadata required to perform Change Data
+///   Feed
+///
+/// ## Context
+/// A note on the `context`. This can be any value the engine wants. This function takes ownership
+/// of the passed arg, but then returns it, so the engine can repeatedly call `visit_cdf_scan_files`
+/// with the same context.
+///
+/// ## Example
+/// ```ignore
+/// let mut context = [my context];
+/// for res in scan_data { // scan data table_changes_scan.scan_data()
+///     let (data, vector, remove_dv) = res?;
+///     context = delta_kernel::table_changes::scan_file::visit_cdf_scan_files(
+///        data.as_ref(),
+///        selection_vector,
+///        context,
+///        my_callback,
+///     )?;
+/// }
+/// ```
+#[allow(unused)]
+pub(crate) fn visit_cdf_scan_files<T>(
+    scan_data: &TableChangesScanData,
+    context: T,
+    callback: CdfScanCallback<T>,
+) -> DeltaResult<T> {
+    let mut visitor = CdfScanFileVisitor {
+        callback,
+        context,
+        selection_vector: &scan_data.selection_vector,
+        remove_dvs: &scan_data.remove_dvs,
+    };
+
+    visitor.visit_rows_of(scan_data.scan_data.as_ref())?;
+    Ok(visitor.context)
+}
+
+/// A visitor that extracts [`CdfScanFile`]s from engine data. Expects data to have the schema
+/// [`cdf_scan_row_schema`].
+#[allow(unused)]
+struct CdfScanFileVisitor<'a, T> {
+    callback: CdfScanCallback<T>,
+    selection_vector: &'a [bool],
+    remove_dvs: &'a Arc<HashMap<String, DvInfo>>,
```

Comment:
why ref arc? should this just be arc?

Reply:
Didn't see a reason to do an entire arc clone in the loop. I changed this to instead have `&'a HashMap<String, DvInfo>`, and we use the arc `as_ref`